Generate solana entrypoint function

dmakarov opened the pull request:
This is an initial implementation of Solana entrypoint function.  The purpose of the function is to implement the interface between Solana run-time/VM and a Move entry function of a specific module. Move entry functions can accept arbitrary parameters. This implementation is limited to calling only entry functions that take no parameters. It extracts the requested entry function name from the instruction_data array passed as input to the entrypoint function from the Solana run-time, and calls the correct entry function. The next step is to implement deserialization of actual parameters and to pass them to entry functions that take arbitrary parameters.

dmakarov committed:
Generate solana entrypoint function

nvjle commented on language/move-native/src/rt.rs:
```diff
@@ -38,7 +38,129 @@ unsafe extern "C" fn vec_cmp_eq(
 }
 
+#[export_name = "move_rt_str_cmp_eq"]
+unsafe fn str_cmp_eq(s1: &str, s2: &str) -> bool {
```
I realize that this patch is a draft, but I'm adding a few minor comments while I familiarize myself with the code.

Can we replace this whole function body with `s1 == s2`? Also, I don't think we need `unsafe` on this one.

dmakarov replied:
Initially, I had `s1 == s2`, but that compared addresses of s1 and s2, not sure why it didn’t expand to element-wise comparison.

nvjle replied:
Does `*s1 == *s2` work?

dmakarov replied:
This seems to work. Thank you.

nvjle commented on language/tools/move-mv-llvm-compiler/src/stackless/llvm.rs:
```diff
@@ -444,6 +468,38 @@ impl Builder {
         }
     }
 
+    /// Get a struct element.
```
This is almost a copy of `field_ref()` (but takes `AnyValue` which is a bit more general). The name is also better for us LLVM types. So perhaps just remove `field_ref` altogether (it isn't used right now).

nvjle commented on language/tools/move-mv-llvm-compiler/src/stackless/module_context.rs:
```diff
@@ -622,4 +624,212 @@ impl<'mm, 'up> ModuleContext<'mm, 'up> {
+        if entry_functions.is_empty()
+            || entry_functions
+                .iter()
+                .any(|fn_env| fn_env.get_name_str() == "<SELF>")
```
Should be able to use `fn_env.is_script()` instead.

dmakarov replied:
Thank you, I’ll use the is_script method instead.

dmakarov replied:
Unfortunately, this method returns true for entry functions also, not only for scripts. I need to exclude only scripts when solana entrypoint is generated.

https://github.com/solana-labs/move/blob/e310bc8c26b58e8d30a7a4267e6f645910141eb4/language/move-model/src/model.rs#L3306-L3310

dmakarov replied:
but I'll use is_script_module()

nvjle commented on language/tools/move-mv-llvm-compiler/src/stackless/module_context.rs:
```diff
@@ -622,4 +624,212 @@ impl<'mm, 'up> ModuleContext<'mm, 'up> {
+        {
+            return;
+        }
+
```
It would be very useful to add a paragraph comment here describing what is happening-- the overall approach of the entry point emitter.

dmakarov replied:
Definitely. Still working on it.

dmakarov replied:
I added a comment, and changed the PR to ready for review. This is still intermediate implementation, and more work is being done, but as a first step I'd like this PR to be considered, so that it doesn't grow too big.

dmakarov committed:
Simplify str_cmp_eq implementation

dmakarov committed:
Remove ref_field method, replaced by getelementptr

dmakarov committed:
Do not generate solana entrypoint for script modules

dmakarov committed:
Clean up entry-point test code

dmakarov committed:
Add a brief description of entrypoint logic

ksolana commented on language/move-native/src/rt.rs:
```diff
@@ -38,7 +38,119 @@ unsafe extern "C" fn vec_cmp_eq(
+    // Number of accounts present
+
+    #[allow(clippy::cast_ptr_alignment)]
+    let num_accounts = *(input.add(offset) as *const u64) as usize;
```
Is the binary format described somewhere, maybe we can add a link to the format here for readability.

dmakarov replied:
Some documentation exists in source code https://github.com/solana-labs/solana/blob/028e22fa232aa410f75bf872408db6ad259c6816/sdk/program/src/lib.rs#L189

nvjle replied:
What I have done in these situations is to write comments for each structure I create which mentions the corresponding runtime type in `move-native`. This makes the code much easier to read. See `rttydesc.rs: define_type_info_global_struct` for an example. I also usually mention that I'm creating and filling up one of those types with the corresponding literal, as well as some running comments on what is being done. Just a suggestion to make the code much easier to understand. That would be especially useful in the entry point generator proper. I do this in my own code specifically so I understand it when revisiting :)

dmakarov replied:
I added comments.

nvjle commented on language/move-native/src/rt_types.rs:
```diff
@@ -209,6 +209,23 @@ impl core::fmt::Debug for MoveAddress {
+pub type SolanaEpoch = u64;
+
+#[repr(transparent)]
+pub struct SolanaPubkey(pub [u8; 32]);
```
32 -> `target_defs::ACCOUNT_ADDRESS_LENGTH`

dmakarov committed:
Suggested changes

nvjle commented on language/tools/move-mv-llvm-compiler/src/stackless/module_context.rs:
```diff
@@ -622,4 +626,216 @@ impl<'mm, 'up> ModuleContext<'mm, 'up> {
+            self.llvm_builder.store(ret, retval);
+            self.llvm_builder.build_br(exit_bb);
+            self.llvm_builder.position_at_end(else_bb);
+        }
```
I think we need one more "default" case emitted that calls `rt_abort` if the user passes an unrecognized entry point name. We do not want to fail silently.

dmakarov replied:
The final `else` block should call `rt_abort`. I'll add that.

nvjle replied:
I wrote some quick tests that use `debug::print` and a `log` line to show different entry point callees actually being correctly called, along with the default/error case. The first calls `entry_point__foo`, the second `entry_point__bar` (those pass), and the third calls an unknown `entry_point__boo` (this should abort, but does not currently).
```
diff --git a/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point01.json b/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point01.json
new file mode 100644
index 000000000..44fae0b50
--- /dev/null
+++ b/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point01.json
@@ -0,0 +1,14 @@
+{
+    "program_id": "DozgQiYtGbdyniV2T74xMdmjZJvYDzoRFFqw7UR5MwPK",
+    "accounts": [
+        {
+            "key": "524HMdYYBy6TAn4dK5vCcjiTmT2sxV6Xoue5EXrz22Ca",
+            "owner": "BPFLoaderUpgradeab1e11111111111111111111111",
+            "is_signer": false,
+            "is_writable": true,
+            "lamports": 1000,
+            "data": [0, 0, 0, 3]
+        }
+    ],
+    "instruction_data": [101, 110, 116, 114, 121, 95, 112, 111, 105, 110, 116, 95, 95, 102, 111, 111]
+}
diff --git a/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point01.move b/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point01.move
new file mode 100644
index 000000000..758e90495
--- /dev/null
+++ b/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point01.move
@@ -0,0 +1,23 @@
+// input entry-point01.json
+// log 17
+
+module 0x10::debug {
+  native public fun print<T>(x: &T);
+}
+
+module 0xa000::entry_point {
+    use 0x10::debug;
+
+    public entry fun bar(): u64 {
+        let rv = 123;
+        debug::print(&rv);
+        rv
+    }
+
+    public entry fun foo(): u64 {
+        let rv = 17;
+        debug::print(&rv);
+        rv
+    }
+}
+
diff --git a/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point02.json b/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point02.json
new file mode 100644
index 000000000..985a0b65f
--- /dev/null
+++ b/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point02.json
@@ -0,0 +1,14 @@
+{
+    "program_id": "DozgQiYtGbdyniV2T74xMdmjZJvYDzoRFFqw7UR5MwPK",
+    "accounts": [
+        {
+            "key": "524HMdYYBy6TAn4dK5vCcjiTmT2sxV6Xoue5EXrz22Ca",
+            "owner": "BPFLoaderUpgradeab1e11111111111111111111111",
+            "is_signer": false,
+            "is_writable": true,
+            "lamports": 1000,
+            "data": [0, 0, 0, 3]
+        }
+    ],
+    "instruction_data": [101, 110, 116, 114, 121, 95, 112, 111, 105, 110, 116, 95, 95, 98, 97, 114]
+}
diff --git a/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point02.move b/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point02.move
new file mode 100644
index 000000000..1609e657c
--- /dev/null
+++ b/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point02.move
@@ -0,0 +1,23 @@
+// input entry-point02.json
+// log 123
+
+module 0x10::debug {
+  native public fun print<T>(x: &T);
+}
+
+module 0xa000::entry_point {
+    use 0x10::debug;
+
+    public entry fun bar(): u64 {
+        let rv = 123;
+        debug::print(&rv);
+        rv
+    }
+
+    public entry fun foo(): u64 {
+        let rv = 17;
+        debug::print(&rv);
+        rv
+    }
+}
+
diff --git a/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point03.json b/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point03.json
new file mode 100644
index 000000000..99105e65d
--- /dev/null
+++ b/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point03.json
@@ -0,0 +1,14 @@
+{
+    "program_id": "DozgQiYtGbdyniV2T74xMdmjZJvYDzoRFFqw7UR5MwPK",
+    "accounts": [
+        {
+            "key": "524HMdYYBy6TAn4dK5vCcjiTmT2sxV6Xoue5EXrz22Ca",
+            "owner": "BPFLoaderUpgradeab1e11111111111111111111111",
+            "is_signer": false,
+            "is_writable": true,
+            "lamports": 1000,
+            "data": [0, 0, 0, 3]
+        }
+    ],
+    "instruction_data": [101, 110, 116, 114, 121, 95, 112, 111, 105, 110, 116, 95, 95, 98, 111, 111]
+}
diff --git a/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point03.move b/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point03.move
new file mode 100644
index 000000000..db731b7b3
--- /dev/null
+++ b/language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point03.move
@@ -0,0 +1,23 @@
+// input entry-point.json
+// abort 8192
+
+module 0x10::debug {
+  native public fun print<T>(x: &T);
+}
+
+module 0xa000::entry_point {
+    use 0x10::debug;
+
+    public entry fun bar(): u64 {
+        let rv = 123;
+        debug::print(&rv);
+        rv
+    }
+
+    public entry fun foo(): u64 {
+        let rv = 17;
+        debug::print(&rv);
+        rv
+    }
+}
+
```

dmakarov replied:
Correct. Currently it exits with exit code 0. I agree that it should abort, and I'll add a call to rt_abort.

nvjle replied:
Could you add the test cases above to your patch?

dmakarov replied:
Will do.

nvjle commented on language/tools/move-mv-llvm-compiler/src/stackless/llvm.rs:
```diff
@@ -284,6 +288,26 @@ impl Module {
+        func: Function,
+        idx: llvm_sys::LLVMAttributeIndex,
+        name: &str,
+        ty: LLVMTypeRef,
```
This should take struct `Type` instead of `LLVMTypeRef`. When that is done, the `clippy::not_unsafe_ptr_arg_deref` can be removed.

nvjle commented on language/tools/move-mv-llvm-compiler/src/stackless/module_context.rs:
```diff
@@ -622,4 +626,216 @@ impl<'mm, 'up> ModuleContext<'mm, 'up> {
+            (3, "nonnull", None),
+        ];
+        let ll_fn_str_cmp = self.llvm_module.add_function("move_rt_str_cmp_eq", ll_fnty);
+        self.llvm_module.add_attributes(ll_fn_str_cmp, &attrs);
```
Ideally, setting up and calling these new runtime functions should be handled the same way as the other existing ones. See `get_runtime_function`, `get_runtime_function_by_name`, enum `RtCall`, and `emit_rtcall_*`. This will also aid readability and make the code more concise.

dmakarov replied:
Should these methods be moved from FunctionContext to ModuleContext?

nvjle replied:
If you mean the list I mentioned above (`get_runtime_functioon`, etc), then I'd say not now (orthogonal). I have done some recent refactoring of the overall `translate.rs` and moving those to `ModuleContext` is probably the right thing to do when more refactoring done (`FunctionContext` hasn't been touched yet and is huge).

dmakarov replied:
Then I'm not sure what your suggestion 

> these new runtime functions should be handled the same way as the other existing ones. See get_runtime_function, get_runtime_function_by_name, enum RtCall, and emit_rtcall_*. 

entails. Should I define similar methods `get_runtime_function` et al. in ModuleContext as well?

nvjle replied:
What I mean is to just follow the existing idiom for runtime functions. For example, in `translate.rs`:
```
pub enum RtCall {
    Abort(mast::TempIndex),
    VecDestroy(mast::TempIndex, mty::Type),
    VecCopy(llvm::AnyValue, llvm::AnyValue, mty::Type),
    VecCmpEq(llvm::AnyValue, llvm::AnyValue, mty::Type),
    VecEmpty(mty::Type),
    StructCmpEq(llvm::AnyValue, llvm::AnyValue, mty::Type),
    /* MY NEW RUNTIME ROUTINE HERE */
}
```
We'd add the new runtime functions there. And likewise, add cases for the other locations:
```
    fn get_runtime_function(&self, rtcall: &RtCall) -> llvm::Function {
        let name = match rtcall {
            RtCall::Abort(..) => "abort",
            RtCall::VecDestroy(..) => "vec_destroy",
            RtCall::VecCopy(..) => "vec_copy",
            RtCall::VecCmpEq(..) => "vec_cmp_eq",
            RtCall::VecEmpty(..) => "vec_empty",
            RtCall::StructCmpEq(..) => "struct_cmp_eq",
            /* MY NEW RUNTIME ROUTINE HERE */
        };
        self.get_runtime_function_by_name(name)
    }
```
And:
```
   fn get_runtime_function_by_name(&self, rtcall_name: &str) -> llvm::Function {
        let fn_name = format!("move_rt_{rtcall_name}");
        let llmod = &self.module_cx.llvm_module;
        let llcx = &self.module_cx.llvm_cx;
        let llfn = llmod.get_named_function(&fn_name);
        if let Some(llfn) = llfn {
            llfn
        } else {
       ...
               "struct_cmp_eq" => {
                    // struct_cmp_eq(type_ve: &MoveType, s1: &AnyValue, s2: &AnyValue) -> bool;
                    let ret_ty = llcx.int_type(1);
                    let tydesc_ty = llcx.int_type(8).ptr_type();
                    let anyval_ty = llcx.int_type(8).ptr_type();
                    let param_tys = &[tydesc_ty, anyval_ty, anyval_ty];
                    let llty = llvm::FunctionType::new(ret_ty, param_tys);
                    let mut attrs = self.mk_pattrs_for_move_type(1);
                    attrs.push((2, "readonly", None));
                    attrs.push((2, "nonnull", None));
                    attrs.push((3, "readonly", None));
                    attrs.push((3, "nonnull", None));
                    (llty, attrs)
                }
                 /* MY NEW RUNTIME ROUTINE HERE */
                
                n => panic!("unknown runtime function {n}"),
            };

            let ll_fn = llmod.add_function(&fn_name, llty);
            llmod.add_attributes(ll_fn, &attrs);
            ll_fn


```
And so forth for the others in the list. This factors out the runtime function declaration and calling to the various routines mentioned.
```

dmakarov replied:
Right, but this logic exists in FunctionContext currently, and not available to emit_solana_entrypoint, because the latter is a method in ModuleContext. Hence, my question was should the code that you pointed out (from translate.rs) be moved into ModuleContext?

nvjle replied:
Oh yes, you are correct. In that case, if it doesn't interrupt your flow, move them to `ModuleContext`. But since I am already doing refactoring now, I could do this as part of that work (and I could then do the factoring mentioned above). Which ever is more convenient for you.

dmakarov replied:
If you're already refactoring this and the methods for handling run-time functions will be moved into ModuleContext, I'll wait for your changes, rebase on that, and will use the rt handling functions to call rt_abort etc. I'll also add the new rt methods (cmp_str, deserialize) to the RtCall enum.  In the meantime I can also work on the entry function parameters (the missing part in this PR), so no rush in making those refactorings.

nvjle replied:
SGTM (I'll commit a minimal refactor soon with just the rt-related factored out).

dmakarov replied:
Thank you!

dmakarov committed:
Add more entry-point tests

nvjle commented on language/tools/move-mv-llvm-compiler/tests/rbpf-tests/entry-point03.move:
```diff
@@ -0,0 +1,22 @@
+// input entry-point03.json
+// abort 8192
```
Incidentally, I just chose 8192 arbitrarily, which will work fine. But it might be best to choose an apropos symbolic value from enum `move-core/types/src/vm_status.rs:vm_status::StatusCode`. This is what I did for overflow/arithmetic errors where we emit the call to `rt_abort` (that one being `vm_status::StatusCode::ARITHMETIC_ERROR`).

dmakarov replied:
Yes, will fix this when I add rt_abort to the default case.

dmakarov committed:
Call abort when unknown entry function is requested in a transaction

nvjle commented on language/tools/move-mv-llvm-compiler/src/stackless/module_context.rs:
```diff
@@ -832,4 +836,214 @@ impl<'mm, 'up> ModuleContext<'mm, 'up> {
+        let slice_init = llcx.const_struct(&[
+            str_literal.ptr(),
+            llcx.const_int_array::<u8>(&slice_len.to_le_bytes())
+                .as_const(),
```
Nit: It's a bit more natural to just use an integer constant here, rather than byte array:
`llvm::Constant::int(llcx.int_type(64), U256::from(slice_len as u128))`

dmakarov replied:
For some reason rust compiler generates the slice constants as a pair of string literal address and a byte array for slice length, but I guess it doesn't matter. I changed it to an int constant.

dmakarov committed:
Use existing idioms to handle new rt functions

dmakarov committed:
Remove entry-point test replaced by entry-point01

dmakarov committed:
Add comments.

dmakarov commented:
I think I addressed all the comments. @nvjle could you, please, take another look when you have a minute? If no further comments, I will merge.

Thank you for reviewing and making good suggestions.

nvjle commented on language/tools/move-mv-llvm-compiler/src/stackless/module_context.rs:
```diff
@@ -772,6 +799,21 @@ impl<'mm, 'up> ModuleContext<'mm, 'up> {
+                    let len_ty = llcx.int_type(64);
+                    let param_tys = &[ptr_ty, len_ty, ptr_ty, len_ty];
+                    let llty = llvm::FunctionType::new(ret_ty, param_tys);
+                    let mut attrs = self.mk_pattrs_for_move_type(1);
```
Either use `mk_pattrs_for_move_type` or below, but not both (on a particular parameter). The former was meant to add attrs for `MoveType`, but in this case, it will work also. That is, you created what amounts to a descriptor with static/constant values. So other than the name not being entirely accurate for this case, `mk_pattrs_for_move_type` is more concise.

dmakarov replied:
Fixed. Thanks for catching this!

nvjle commented on language/tools/move-mv-llvm-compiler/src/stackless/module_context.rs:
```diff
@@ -832,4 +874,198 @@ impl<'mm, 'up> ModuleContext<'mm, 'up> {
+            params,
+            ll_fn_solana_entrypoint.get_param(0).as_any_value(),
+        ));
+        self.llvm_builder.call(ll_fn_deserialize, &input);
```
Why not setup and invoke like the other rt calls? But fine for now.

dmakarov replied:
I factored out the emission of deserialize call into a separate function.

nvjle reviewed (approved):
Two minor comments, otherwise LGTM.

dmakarov committed:
Factor deserialize call emit into a separate function

nvjle commented:
Final commit LGTM, ready to land.
